### DISCUSSION TITLE
Vercel Rewrites in Docusaurus App

### DISCUSSION THREAD

**Discussion by brian-tenovos** [vercel/community#2548] [Help] [answered]
### Summary

I am attempting to use the vercel.json file to perform rewrites within a Docusaurus app.  I am wanting to capture all query params that exist on any page of the /docs directory and route the user to a specific page based on the values of the query strings.  

PROBLEM: Every time I add query params to one of my requests I get a 404.

### Additional information

Here is my vercel.json file.  I would repeat this for each path I need to rewrite I suppose, without a better option.

`{
    "rewrites": [
        {
            "source": "/docs",
            "has": [
                {
                    "type": "query",
                    "key": "context",
                    "value": "search"
                }
            ],
            "destination": "/docs/search"
        }
    ]
}`

**Comment by amyegan** [comment]
Hi @brian-tenovos. Do you have a minimal reproduction or public repo that I can use to step through the error and get more details?

**Reply by brian-tenovos** [reply]
Hi @amyegan ,

The site can be viewed from here: https://tnvs-docs-app.vercel.app/

The vercel config right now looks like this:

`{
    "rewrites": [
        {
            "source": "/docs",
            "has": [
                {
                    "type": "query",
                    "key": "context",
                    "value": "watermark"
                }
            ],
            "destination": "/docs/watermark"
        }
    ]
}`

So, https://tnvs-docs-app.vercel.app/docs/?context=watermark should navigate users to https://tnvs-docs-app.vercel.app/docs/watermark, however, instead a Page Not Found page is displayed and the query params stay in the url.

**Reply by amyegan** [reply]
I was able to replicate what you're seeing in my own copy of the [Vercel Docusaurus template](https://vercel.com/templates/react/docusaurus-2). This seems to be because of the way Docusaurus handles routing and filesystem priority when handling rewrites. Docusaurus may have it's own way of handling rewrites that would work for your site, or you can switch to redirects via [vercel.json config](https://vercel.com/docs/concepts/projects/project-configuration#redirects) or [Middleware](https://vercel.com/docs/concepts/functions/edge-middleware/middleware-api).

**Reply by brian-tenovos** [reply]
Hi Amy, so a redirect instead of a rewrite in the vercel.json should work?  I’ve tried that as well without any luck.

**Reply by amyegan** [reply, marked as answer]
Try with Middleware instead. This is what worked in my test app:

```javascript
/* middleware.js */

// config with custom matcher
export const config = {
  matcher: '/docs',
};

export default function middleware(request) {
  // get the query params
  const url = new URL(request.url);
  const params = url.searchParams;

  // if 'context' is available and value is 'basic', redirect
  if(params.get("context") == "basic") {
    return Response.redirect(new URL('/docs/category/tutorial---basics', request.url));
  }
}
```

**Reply by brian-tenovos** [reply]
Thank you Amy!

**Comment by brian-tenovos** [comment]
@amyegan Is any additional information needed?
